realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -0,0 +1,269 @@
+            // This is the first instance in current thread, increase the global count.
+            refAndCount.globalCount++;
+        }
+        refAndCount.localCount.set(refCount + 1);
+
+        return refAndCount.localRealm.get();
+    }
+
+    /**
+     * Releases a given {@link Realm} or {@link DynamicRealm} from cache. The instance won't be closed by this function.
+     *
+     * @param realm Realm instance to be released from cache.
+     * @return local reference count to the type Realm after releasing. {@code -1} if the Realm instance is not in the
+     *         cache.
+     */
+    static synchronized int release(BaseRealm realm) {
+        String canonicalPath = realm.getPath();
+        RealmCache cache = cachesMap.get(canonicalPath);
+        Integer refCount = null;
+        RefAndCount refAndCount = null;
+
+        if (cache != null) {
+            refAndCount = cache.refAndCountMap.get(RealmCacheType.valueOf(realm.getClass()));
+            refCount = refAndCount.localCount.get();
+        }
+        if (refCount == null) {
+            refCount = 0;
+        }
+
+        if (refCount <= 0) {
+            RealmLog.w("Realm " + canonicalPath + " is not in the cache anymore.");
+            return -1;
+        }
+
+
+        // Decrease the local counter.
+        refCount -= 1;
+
+        if (refCount == 0) {
+            // The last instance in this thread.
+            // Clear local ref & counter
+            refAndCount.localCount.set(null);
+            refAndCount.localRealm.set(null);
+
+            // Clear global counter
+            refAndCount.globalCount--;
+            if (refAndCount.globalCount < 0) {
+                // Should never happen.
+                throw new IllegalStateException("Global reference counter of Realm" + canonicalPath +
+                        " gets corrupted.");
+            }
+
+            // Clear the column indices cache if needed
+            if (realm instanceof Realm && refAndCount.globalCount == 0) {
+                // All typed Realm instances of this file are cleared from cache
+                cache.typedColumnIndices = null;
+            }
+
+            int totalRefCount = 0;
+            for (RealmCacheType type : RealmCacheType.values()) {
+                totalRefCount += cache.refAndCountMap.get(type).globalCount;
+            }
+            // No more instance of typed Realm and dynamic Realm. Remove the configuration from cache.
+            if (totalRefCount == 0) {
+                cachesMap.remove(canonicalPath);
+            }
+        } else {
+            refAndCount.localCount.set(refCount);
+        }
+
+        return refCount;
+    }
+
+    /**
+     * Make sure that the new configuration doesn't clash with any existing configurations for the
+     * Realm.
+     *
+     * @throws IllegalArgumentException If the new configuration isn't valid.
+     */
+    private static void validateAgainstExistingConfigurations(RealmConfiguration newConfiguration,
```

Comment:
In the previous incarnation you could have multiple configurations pointing to the same file. Now there only seem to be one allowed. I am not sure we had a use case yet for the other implementation so it is probably fine, but at least renaming this to `validateAgainstExistingConfiguration`